full-service/src/service/transaction.rs

```diff
@@ -1674,4 +1703,345 @@ mod tests {
+            .map(|(t, _)| Txo::get(&t.id, service.get_pooled_conn().unwrap().deref_mut()).unwrap())
+            .collect::<Vec<Txo>>();
+        assert_eq!(change.len(), 1);
+        assert_eq!(change[0].value as u64, 58 * MOB - Mob::MINIMUM_FEE);
+
+        let inputs = transaction_txos
+            .inputs
+            .iter()
+            .map(|t| Txo::get(&t.id, service.get_pooled_conn().unwrap().deref_mut()).unwrap())
+            .collect::<Vec<Txo>>();
+        assert_eq!(inputs.len(), 1);
+        assert_eq!(inputs[0].value as u64, 100 * MOB);
+
+        // Verify balance for Alice's subaddress = original balance - fee - txo_value
+        // NOTE: This confirms that the change went back to Alice's subaddress, as it
+        // should have, rather than the default change subaddress
+        let balance = service
+            .get_balance_for_address(&alice_subaddress.public_address_b58)
+            .unwrap();
+        let balance_pmob = balance.get(&Mob::ID).unwrap();
+        assert_eq!(balance_pmob.unspent, (58 * MOB - Mob::MINIMUM_FEE) as u128);
+
+        // Bob's balance should be = output_txo_value
+        let bob_balance = service
+            .get_balance_for_address(&bob_subaddress.public_address_b58)
+            .unwrap();
+        let bob_balance_pmob = bob_balance.get(&Mob::ID).unwrap();
+        assert_eq!(bob_balance_pmob.unspent, 42 * MOB as u128);
+
+        // Decrypt the memo from the transaction txo (from Alice to Bob) and verify.
+        let alice_to_bob_txo = &tx_proposal.payload_txos[0].tx_out;
+
+        let shared_secret = get_tx_out_shared_secret(
+            exchange_account_key.view_private_key(),
+            &RistrettoPublic::try_from(&alice_to_bob_txo.public_key).unwrap(),
+        );
+
+        let memo = alice_to_bob_txo.decrypt_memo(&shared_secret);
+        let authenticated_sender_memo = AuthenticatedSenderMemo::from(memo.get_memo_data());
+        let validation = authenticated_sender_memo.validate(
+            &exchange_account_key.subaddress(alice_subaddress.subaddress_index as u64),
+            &exchange_account_key.subaddress_view_private(bob_subaddress.subaddress_index as u64),
+            &alice_to_bob_txo.public_key,
+        );
+        assert!(bool::from(validation));
+
+        // Decrypt the memo from the change txo (From Alice back to Alice) and verify
+        let alice_to_bob_change_txo = &tx_proposal.change_txos[0].tx_out;
+
+        let change_shared_secret = get_tx_out_shared_secret(
+            exchange_account_key.view_private_key(),
+            &RistrettoPublic::try_from(&alice_to_bob_change_txo.public_key).unwrap(),
+        );
+
+        // FIXME: This is ideally how this works once we update the
+        // `transaction_builder.add_change_output` to accept a non-change
+        // subaddress. For now, we're just decrypting the memo as an authenticated
+        // sender memo.
+        let change_memo = alice_to_bob_change_txo.decrypt_memo(&change_shared_secret);
```

Comment:
Note: ⚠️ This is where the change memo logic is being tested, and flagged how it should be tested if DestinationMemo is properly constructed

Reply:
Destination memo is now being properly constructed, so resolving this comment